sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala

```diff
@@ -120,6 +120,30 @@ case class InlineCTE(alwaysInline: Boolean = false) extends Rule[LogicalPlan] {
           }
         }
     }
+    // If a CTE ref count is 0, the other CTE that is referencing it should also have
```

Comment:
I'm a bit concerned about changing the ref count directly. I think the ref count is correct, the problem is in `CheckAnalysis` when we decide which CTE relations need to be checked ahead. The current condition is simply `refCount == 0`, but we should also check CTE relations referenced by CTE relations that has 0 refCount.

Reply:
I agree with @cloud-fan. Probably you just call `cleanCTEMap()` in `CheckAnalysis` too.

Reply:
Tried to address this comment.